packages/composites/gm-token/src/transport/price.ts

```diff
@@ -0,0 +1,238 @@
+
+    const assets = [index, long, short]
+
+    const providerDataRequestedUnixMs = Date.now()
+
+    const [indexPrices, longPrices, shortPrices] = await this.fetchPrices(
+      assets,
+      providerDataRequestedUnixMs,
+    )
+
+    const indexToken = tokenAddresses.arbitrum[index]
+    const longToken = tokenAddresses.arbitrum[long]
+    const shortToken = tokenAddresses.arbitrum[short]
+
+    const tokenPriceContractParams = [
+      this.settings.DATASTORE_CONTRACT_ADDRESS,
+      [market, indexToken, longToken, shortToken],
+      [indexPrices.ask, indexPrices.bid],
+      [longPrices.ask, longPrices.bid],
+      [shortPrices.ask, shortPrices.bid],
+      utils.keccak256(this.abiEncoder.encode(['string'], [this.settings.PNL_FACTOR_TYPE])),
+    ]
+
+    // Prices have a spread from min to max. The last param (maximize-true/false) decides whether to maximize the market token price
+    // or not. We get both values and return the median.
+    const [[maximizedValue], [minimizedValue]] = await Promise.all([
+      this.readerContract.getMarketTokenPrice(...tokenPriceContractParams, true),
+      this.readerContract.getMarketTokenPrice(...tokenPriceContractParams, false),
+    ])
+
+    const maximizedPrice = Number(utils.formatUnits(maximizedValue, SIGNED_PRICE_DECIMALS))
+    const minimizedPrice = Number(utils.formatUnits(minimizedValue, SIGNED_PRICE_DECIMALS))
+    const result = median([minimizedPrice, maximizedPrice])
+
+    return {
+      data: {
+        result,
+      },
+      statusCode: 200,
+      result,
+      timestamps: {
+        providerDataRequestedUnixMs,
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+  }
+
+  // Fetches the lwba price info from multiple source EAs, calculates the median for bids and asks per asset and fixes the price precision
+  private async fetchPrices(assets: string[], dataRequestedTimestamp: number) {
+    // priceData holds raw bid/ask values per asset from source EAs response
+    const priceData = {} as PriceData
+    assets.forEach((a) => {
+      priceData[a] = { bids: [], asks: [] }
+    })
+
+    const sources = [
+      this.settings.TIINGO_ADAPTER_URL,
+      this.settings.COINMETRICS_ADAPTER_URL,
+      this.settings.NCFX_ADAPTER_URL,
+    ]
+
+    const promises = []
+
+    for (let i = 0; i < sources.length; i++) {
+      const source = sources[i]
+
+      const assetPromises = assets.map(async (asset) => {
+        const requestConfig = {
+          url: source,
+          method: 'POST',
+          data: {
+            data: {
+              endpoint: 'crypto-lwba',
+              base: this.unwrapAsset(asset),
+              quote: 'USD',
+            },
+          },
+        }
+
+        // try/catch is mostly needed in a case if one of source EAs fails to return a response,
+        // we will still get and calculate the median price based on responses of remaining EAs
```

Comment:
Have we discussed if this is the behaviour we want?

My concern is that if 2 of the 3 EAs go down, we could potentially just be relying on their price without noticing. This might be fine, but I want to check this is a risk we've considered!

Reply:
Optional: Is there any way to return metadata in the end response which tells us which EAs returned data that made up the medianised price?

Reply:
This was not discussed yet. I will clarify this and if we can skip failing EAs i will change the response to include the list of sources that provided price info.